Remove synchronous postgres_backend

arssher opened the pull request:
Remove sync postgres_backend_async, tidy up its split usage.

- Use postgres_backend_async throughout safekeeper.
- Use framed.rs in postgres_backend_async, similar to tokio_util::codec::Framed
  but with slightly more efficient split. Now IO functions are also cancellation
  safe. Also, there is no allocation in each message read anymore, and data in
  read messages still directly point to input buffer without copies.
- In both safekeeper COPY streams, do read-write from the same thread/task with
  select! for easier error handling.
- Tidy up finishing CopyBoth streams in safekeeper sending and receiving WAL
  -- join split parts back catching errors from them before returning.

Initially I hoped to do that read-write without split at all, through polling
IO:
https://github.com/neondatabase/neon/pull/3522
However that turned out to be more complicated than I initially expected
due to 1) borrow checking and 2) anon Future types. 1) required Rc<Refcell<...>>
which is Send construct just to satisfy the checker; 2) can be workaround with
transmute. But this is so messy that I decided to leave split.

proxy stream.rs is adapted with minimal changes. It also benefits from framed.rs
improvements described above.

TODO:
- [x] rebase
- [x] fix proxy
- [x] fix unit tests
- [x] better error logging
- [x] probably simplify send_wal.rs

arssher commented:
There is TODO list above, but otherwise it is ready for review; main tests pass. CC @petuhovskiy

petuhovskiy reviewed (commented):
Looked at everything but `safekeeper/src/send_wal.rs`, going to take another look tomorrow and probably post more minor comments, but so far looks good.

Found some minor doc issues which are not worth making github comment not to pollute discussion:
- There were several typos in the comments, I think it should be possible to run some sort of grammar linter to fix them quickly (e.g. https://github.com/neondatabase/neon/pull/1818#issuecomment-1140273811)
- rustdoc comments (`///`) are missed in several places and there are usual `//` instead

Also:
- `write_message_noflush -> write_message` and `write_message -> write_message_flush` can be confusing or error-prone, inverting the `write_message` semantics looks strange
- Calling `unwrap()` and `expect()` in functions which are returning `Result<>` doesn't feel right IMO (except when checked with `is_ok` on previous line), even if we are sure that it won't panic _normally_

arssher commented:
Simplifed `send_wal.rs` by moving to socket split based approach with select!, like receive_wal.rs.

arssher commented:
@funbringer Changes to proxy are not too invasive, but what is the basic way of testing this? I replaced guts of stream.rs and switched mgmt.rs to async.

arssher commented:
> There were several typos in the comments, 
Run codespell. Most of suggestions were s/crate/create, but I fixed the rest :)

> rustdoc comments (///) are missed in several places
Tried to look through most of files and added some, but point if you see more.

> write_message_noflush -> write_message and write_message -> write_message_flush can be confusing or error-prone, 
Yeah, I already bumped into this a couple of times. However, postgres_backend_async already has different meaning, so we break semantic anyway; and I'm much more proficient in safekeeper code than in pageserver, so adoped what was already there.

> Calling unwrap() and expect() in functions which are returning Result<> doesn't feel right IMO 

Removed most of these. A couple left, but they are mostly independent to this patch, should be fixed separately.

arssher commented on libs/utils/src/postgres_backend_async.rs:
```diff
@@ -152,18 +151,86 @@ impl AsyncRead for Stream {
+    }
+
+    async fn shutdown(&mut self) -> io::Result<()> {
+        match self {
```
makes sense, changed that way

arssher commented on safekeeper/src/receive_wal.rs:
```diff
@@ -2,216 +2,289 @@
-                    }
+    // Forward all messages to WalAcceptor
+    let res = read_network_loop(pgb_reader, msg_tx, next_msg).await;
+    // Unregister connection. XXX this is much more suitable for Drop, but async
```
I don't follow. `on_compute_disconnect` is needed to track number of connected computes; we should call it when stream ends for whatever reason.

petuhovskiy replied:
Yes, we should always call it, that's exactly what I wanted to ask. Is it possible that async task gets dropped before we call `on_compute_disconnect`?

arssher replied:
Ah. No, this is not possible, there is no high level cancellation.

arssher replied:
Hm, select! can kill it on the second thought, nice catch.

arssher replied:
Reworked with Drop which calls spawn_blocking. Ugly, but should be fine.

petuhovskiy commented on libs/pq_proto/src/framed.rs:
```diff
@@ -0,0 +1,256 @@
+    }
+
+    pub async fn shutdown(&mut self) -> Result<(), io::Error> {
+        self.stream.shutdown().await
```
Let's call `flush()` here before shutdown?

arssher replied:
forgot that, added

petuhovskiy commented on libs/pq_proto/src/framed.rs:
```diff
@@ -0,0 +1,256 @@
+    }
+
+    pub async fn shutdown(&mut self) -> Result<(), io::Error> {
+        self.stream.shutdown().await
```
Let's call `flush()` here before shutdown?

petuhovskiy commented on libs/pq_proto/src/lib.rs:
```diff
@@ -179,255 +174,205 @@ pub struct FeExecuteMessage {
-                    .map_err(ConnectionError::Socket)?;
-                Bytes::from(buffer)
-            };
+        // lenth field includes itself, but not message type.
```
```suggestion
        // length field includes itself, but not message type.
```

petuhovskiy commented on libs/pq_proto/src/lib.rs:
```diff
@@ -179,255 +174,205 @@ pub struct FeExecuteMessage {
-                Err(e) => return Err(ConnectionError::Socket(e)),
-            };
+        if buf.len() < 4 {
+            let to_read = 5 - buf.len();
```
Deserves a comment on why it's `5`

arssher replied:
it should be 4 actually, a copypaste typo :) added comment

petuhovskiy commented on libs/utils/src/postgres_backend_async.rs:
```diff
@@ -355,17 +434,77 @@ impl PostgresBackend {
+                Ok(reader)
+            }
+            MaybeWriteOnly::WriteOnly(_) => {
+                anyhow::bail!("PostgresBackend is already split")
```
Note that `self` can be left in `Broken` state here. It seems ok, because it worked the same in `start_tls()` before.

arssher replied:
Yes. But this is one of the errors which should never happen unless there is a programmer error in the server; can't do much then anyway.

petuhovskiy commented on libs/utils/src/postgres_backend_async.rs:
```diff
@@ -570,16 +713,19 @@ impl<'a> AsyncWrite for CopyDataWriter<'a> {
-        this.pgb.write_message(&BeMessage::CopyData(buf))?;
+        this.pgb
+            .write_message(&BeMessage::CopyData(buf))
+            // write_message only writes to buffer, so can fail iff message is
+            // invaid, but CopyData can't be invalid.
```
```suggestion
            // write_message only writes to the buffer, so it can fail
            // if the message is invalid, but CopyData can't be invalid.
```

arssher replied:
iff is intentional here, otherwise applied

arssher replied:
https://en.wikipedia.org/wiki/If_and_only_if

petuhovskiy commented on pageserver/src/tenant/timeline/walreceiver/walreceiver_connection.rs:
```diff
@@ -434,9 +434,7 @@ fn ignore_expected_errors(pg_error: postgres::Error) -> anyhow::Result<postgres:
     {
         return Ok(pg_error);
     } else if let Some(db_error) = pg_error.as_db_error() {
-        if db_error.code() == &SqlState::CONNECTION_FAILURE
```
Why `db_error.code()` is not checked anymore?

arssher replied:
Previously we sent io error in this case:
```
                            return Err(QueryError::from(io::Error::new(
                                io::ErrorKind::ConnectionAborted,
                                format!("end streaming to {:?}", spg.appname),
                            )));
```

That's not really an io error, so I changed to throw anyhow instead, which gets transformed into PG kisses and hugs `XXOOO` internal server error, not `CONNECTION_FAILURE`. That's perhaps not the most appropriate errcode as well, but better typing here can be made separately, it needs changed in postgres_backend API.

arssher replied:
> If someone will write an error message
That's extremely unlikely, but I've found there is a SUCCESSFUL_COMPLETION just for this case.

LizardWizzard replied:
> That's extremely unlikely

The words are not unusual, so it is unlikely, but can happen. 

Thanks for bringing back the check

petuhovskiy commented on safekeeper/src/http/routes.rs:
```diff
@@ -181,16 +184,54 @@ async fn timeline_create_handler(mut request: Request<Body>) -> Result<Response<
+// from safekeeper in isolation, e.g.
+// pg_receivewal -v -d "host=localhost port=5454 options='-c tenant_id=deadbeefdeadbeefdeadbeefdeadbeef timeline_id=deadbeefdeadbeefdeadbeefdeadbeef'" -D ~/tmp/tmp/tmp
+// (hacking pg_receivewal startpos is currently needed though to make pg_receivewal work)
+async fn create_fake_timeline_handler(_request: Request<Body>) -> Result<Response<Body>, ApiError> {
```
I don't really like `create_fake_timeline_handler`:
- `pg_version`, `lsn`, `tenant_id/timeline_id`, etc. are fixed
- we already have another API (`START_WAL_PUSH`) to create timelines and write WAL; it seems that we can craft an external tool to connect to any safekeeper and write anything we want

Should we include it in this PR?

arssher replied:
Yeah, I agree it is really ad hoc. However, it is useful because it allows to test `START_REPLICATION` in isolation, I added it when reworked `send_wal.rs` while `receive_wal.rs` was still broken. It might be useful for other similar purposes. But I don't insist, can remove it.

petuhovskiy replied:
I'd removed it. I'm also thinking about removing json_ctrl from time to time :)

arssher replied:
ok

> I'm also thinking about removing json_ctrl

+1

arssher commented on libs/pq_proto/src/framed.rs:
```diff
@@ -0,0 +1,253 @@
+        // If we can't build a frame yet, try to read more data and try again.
+        // Make sure we've got room for at least one byte to read to ensure
+        // that we don't get a spurious 0 that looks like EOF.
+        read_buf.reserve(1);
```
This piece was actually directly copied from
https://docs.rs/tokio-util/latest/src/tokio_util/codec/framed_impl.rs.html#220

AFAIU partially this `reserve(1)` is more like guard against badly behaving codec; often codec requests chunk big enough to get the whole packet and this is no op. Sometimes it might not know it (not even first bytes of packet were received), and buffer can be indeed full. However, even in this case this doesn't mean inefficient realloc of +1 bytes size; `BytesMut` will first try to shift data left, occuping already released space in the buffer, and most likely will succeed in that if we Drop old messages. That's my general understanding, I haven't studied `bytes` in detail though; but I'm also curious about that.

> Also, read_buf has a strange [doc](https://docs.rs/tokio/latest/tokio/io/trait.AsyncReadExt.html#return-1)

Some streams, in theory, e.g. something in memory can produce EOF and then again data if something changes. This is definitely not the case for `TcpStream`, but this is generic func.

petuhovskiy commented on safekeeper/src/wal_storage.rs:
```diff
@@ -502,6 +502,11 @@ impl WalReader {
         })
     }
 
+    pub async fn fake_read(&mut self) -> Result<usize> {
```
Is it used anywhere or can be removed?

arssher replied:
Forgot to remove after debugging, done.

petuhovskiy commented on safekeeper/src/receive_wal.rs:
```diff
@@ -2,216 +2,289 @@
+        }
+    };
+    // Now that we have timeline, register the connection.
+    tli.on_compute_connect().await?;
```
Maybe it's better to call `on_compute_connect` and `on_compute_disconnect` inside `WalAcceptor::run`? Because currently `on_compute_...` is not synchronized with `tli.process_msg()` calls as they are processed in separate threads.

arssher replied:
It is actually synchronized, as WalAcceptor does nothing until first message is received, which happens after `on_compute_connect`. I'd like to register connections in global map (to e.g. be able to fetch them with GET), and so place of registration should know peer SocketAddr, in that sense networking thread is closer. But this is not really important, either thread is ok.

Ideally I'd prefer to do that directly in `handle_start_wal_push`, but without explicit timeline creation we don't have a timeline until first message arrives, so I just register as soon as it becomes possible.

petuhovskiy replied:
`on_compute_disconnect` can be called before we finish processing all messages inside `WalAcceptor::run`. That's not a big deal, but it can be confusing if we will track this.

`WalAcceptor` right now looks like a nice connection object, it has I/O channels, no messages are processed until it's spawned and no messages are processed after it was dropped. It seems logical to call `on_compute_connect` in the constructor and `on_compute_disconnect` in the destructor, that's my main point.

arssher replied:
Haven't thought of that, agree.

arssher replied:
done it that way

arssher commented on safekeeper/src/receive_wal.rs:
```diff
@@ -2,216 +2,289 @@
+        // sends, so this avoids deadlocks.
+        let mut pgb_reader = pgb.split().context("START_WAL_PUSH split")?;
+        let peer_addr = *pgb.get_peer_addr();
+        let res = tokio::select! {
```
No. The very idea is that we stop as soon as any of parts stops; this allows to easily retrieve error (not choosing among two, trying to guess which was the initial cause) and avoid synchronizing.

LizardWizzard replied:
Looks like [try_join_all](https://docs.rs/futures/latest/futures/future/fn.try_join_all.html) can be helpful

arssher replied:
It would work too, but select! here is enough.

arssher replied:
It has just different semantics. Once any task finished, whether with Ok or error, we need to stop polling. So on the second though it doesn't fit here.

LizardWizzard replied:
This exactly matches try_join_all doc: 

> If any future returns an error then all other futures will be canceled and an error will be returned immediately.

Am I missing something?

arssher replied:
We need to stop even if one future returned Ok.

petuhovskiy commented on safekeeper/src/receive_wal.rs:
```diff
@@ -2,204 +2,286 @@
+            // todo: add read|write .context to these errors
+            r = read_network(self.ttid, &mut pgb_reader, peer_addr, msg_tx, &mut acceptor_handle, msg_rx, reply_tx) => r,
+            r = write_network(pgb, reply_rx) => r,
+        };
```
I looked again, `write_network` will probably finish because of the closed channel, so no problem here.

arssher replied:
yes

arssher commented:
Also reworked error handling, now double logging of closing connections on safekeepers is gone, and generally it is more accurate.

petuhovskiy commented on safekeeper/src/lib.rs:
```diff
@@ -95,3 +96,11 @@ impl SafeKeeperConf {
     }
 }
+
+// Get unique thread id (Rust internal), with ThreadId removed for shorter printing
```
```suggestion
/// Get unique thread id (Rust internal), with ThreadId removed for shorter printing
```

petuhovskiy commented on safekeeper/src/receive_wal.rs:
```diff
@@ -2,204 +2,285 @@
+        pgb: &mut PostgresBackend,
+    ) -> Result<(), QueryError> {
+        if let Err(end) = self.handle_start_wal_push_guts(pgb).await {
+            pgb.handle_copy_stream_end(end).await;
```
Let's add a comment that the error is logged inside `handle_copy_stream_end`, it's not obvious.

arssher replied:
done

arssher replied:
weirdly gh doesn't show updated diff in this chunk

LizardWizzard reviewed (commented):
Looked only at framed.rs. Will continue

funbringer reviewed (commented):
Posting the results of my superficial review. Will come back later with an alternative to the parsing machinery.

arssher commented on proxy/src/stream.rs:
```diff
@@ -51,17 +46,30 @@ impl<S: AsyncRead + Unpin> PqStream<S> {
             .await
             .map_err(ConnectionError::into_io_error)?
             .ok_or_else(err_connection)?;
 
         match msg {
             FeMessage::StartupPacket(packet) => Ok(packet),
-            _ => panic!("unreachable state"),
+            msg => Err(io::Error::new(
+                io::ErrorKind::Other,
+                format!("expected startup message, got {:?}", msg),
+            )),
```
Makes sense.

arssher replied:
Done that, please check again.

funbringer commented on safekeeper/src/lib.rs:
```diff
@@ -95,3 +96,11 @@ impl SafeKeeperConf {
     }
 }
+
+/// Get unique thread id (Rust internal), with ThreadId removed for shorter printing
+pub fn get_tid() -> u64 {
+    let tids = format!("{:?}", thread::current().id());
+    let r = Regex::new(r"ThreadId\((\d+)\)").unwrap();
+    let caps = r.captures(&tids).unwrap();
+    caps.get(1).unwrap().as_str().parse().unwrap()
+}
```
https://docs.rs/nix/latest/nix/unistd/fn.getpid.html

arssher replied:
Yeah, let's remove this shame. We hadn't nix dependency when this was added.

LizardWizzard reviewed (changes requested):
Thank you for working on cleaning this up. I like the direction where its going, but IMO it still needs more work to be mergeable.

Note: I havent really looked at safekeeper changes

LizardWizzard commented on safekeeper/src/lib.rs:
```diff
@@ -95,3 +96,11 @@ impl SafeKeeperConf {
 }
+
+/// Get unique thread id (Rust internal), with ThreadId removed for shorter printing
+pub fn get_tid() -> u64 {
```
Hm, or I'm wrong?  Will it cache the value instead? Still I think if you need the pid the better option would be to use corresponding functions from nix crate ([gettid](https://docs.rs/nix/latest/nix/unistd/fn.gettid.html)?) since we already have it as a dependency.

arssher replied:
fixed, see discussion aove

funbringer commented on libs/pq_proto/src/lib.rs:
```diff
@@ -191,255 +186,206 @@ pub struct FeExecuteMessage {
+    pub fn parse(buf: &mut BytesMut) -> Result<Option<FeMessage>, ProtocolError> {
+        // Every message contains message type byte and 4 bytes len; can't do
+        // much without them.
+        if buf.len() < 5 {
```
I don't think such constant will help here. It's "1 + 4" specifically, not 5. The way we dance around with "1+4", "len + 1" and such doesn't look good to me, though. Reader with read_u8 and then read_u32 etc would look much better.

arssher replied:
> I don't think such constant will help here. It's "1 + 4" specifically, not 5.

+1

> Reader with read_u8 and then read_u32 etc would look much better.

You'd still need to do 'consume' in the end with something like 'len + 1', seems to some degree unavoidable.

funbringer commented on libs/remote_storage/src/lib.rs:
```diff
@@ -111,7 +111,7 @@ pub trait RemoteStorage: Send + Sync + 'static {
 
 pub struct Download {
-    pub download_stream: Pin<Box<dyn io::AsyncRead + Unpin + Send>>,
+    pub download_stream: Pin<Box<dyn io::AsyncRead + Unpin + Send + Sync>>,
```
I suppose it's referenced somewhere. &Send is !Sync.

arssher replied:
Yes. Safekeeper walsender became async, `WalReader` containing such stream (due to taking WAL from s3) is now held across awaits and so it has to be Sync since futures can be moved between threads of the runtime.

arssher commented on libs/pq_proto/src/lib.rs:
```diff
@@ -191,255 +186,206 @@ pub struct FeExecuteMessage {
+        // We shouldn't advance `buf` as probably full message is not there yet,
+        // so can't directly use Bytes::get_u32 etc.
+        let len = (&buf[0..4]).read_u32::<BigEndian>().unwrap() as usize;
+        if !(8..=MAX_STARTUP_PACKET_LENGTH).contains(&len) {
```
I don't, clippy insists on it. I disagree with it.

funbringer replied:
Then we should probably disable it project-wide.

arssher replied:
Added `manual-range-contains` to exclusions, reverted.

arssher commented on libs/pq_proto/src/framed.rs:
```diff
@@ -0,0 +1,265 @@
+//! the async stream based on (and buffered with) BytesMut. All functions are
+//! cancellation safe.
+//!
+//! It is similar to what tokio_util::codec::Framed with appropriate codec
```
Yes, the only supposed way to use split Framed is split on Stream/Sink:
https://docs.rs/tokio-util/latest/tokio_util/codec/struct.Framed.html#method.from_parts

See split.rs and bilock.rs in futures-util.

Funnily, there is even a crate doing bilock with allocation:
https://github.com/paritytech/bilock

LizardWizzard replied:
The question still remains. Why we cannot use existing `tokio_util::codec::Framed`?

arssher replied:
Because I don't like allocation in the hot path of reading/writing each message in split case (which is main for safekeepers). Additionally, though less important, as this comment says having same error for all io framed functions is better avoided. Also we don't really need Stream/Sink abstractions here (apart from mental/potential perf overhead it creates friction like returning `Option<Result<...>>` on next message, while `Result<Option<...>>` is easier to handle).

LizardWizzard replied:
Can you please point to the place where these allocations happen?

>  as this comment says having same error for all io framed functions is better avoided

> Will an enum with different error variants be a decent way to handle that?

I'm not sure I understand which methods are problematic. Can you elaborate on that?

> returning Option<Result<...>> on next message, while Result<Option<...>> is easier to handle).

See [Option::transpose](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.transpose). Will it help? 

My concern is the code that in some places duplicates code from library (futures, tokio), and I still do not understand why do we need to invent this ourselves.

arssher replied:
https://docs.rs/futures-util/latest/src/futures_util/lock/bilock.rs.html#107

In the final version of the patch in safekeeper we read-write only from the same task/thread (currently; in theory this might also change), so on the first glance this might be not an issue . I still don't think we'd have sensible simplicity win with tokio_util Framed though; currently all related logic is last ~50 LOC in `framed.rs`. tokio_util Framed itself is more lenghty and complicated in ways we don't need. Finally, @funbringer expressed eager desire to have .await's in protocol parsing functions, that's not compatible with tokio_util Framed as well.

> I'm not sure I understand which methods are problematic.

read_message, write_message, flush all return different errors reflecting what they do. Having function declared as returning variants it never actually returns makes code harder to read.

> See [Option::transpose](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.transpose). 

Nice, I found [thread](https://users.rust-lang.org/t/convenience-method-for-flipping-option-result-to-result-option/13695/24) when googled that, but haven't scrolled to the bottom. This is minor though.

LizardWizzard replied:
> https://docs.rs/futures-util/latest/src/futures_util/lock/bilock.rs.html#107

Ok, makes sense, please add this to the comment if we will keep this code. From the other side, in tokio instead of keeping wakers around it spins when trying to acquire the lock https://docs.rs/tokio/latest/src/tokio/io/split.rs.html#145. Not sure if this is a problem though.

> I still don't think we'd have sensible simplicity win with tokio_util Framed though; currently all related logic is last ~50 LOC in framed.rs. tokio_util Framed itself is more lenghty and complicated in ways we don't need.

I wont insist. To me tokio's framed can allow for better composition with other tokio related utilities. It may be reasonable to keep the scope smaller. Lets see if we can do better with @funbringer's approach

arssher replied:
added that link to the comment

funbringer commented on libs/utils/tests/postgres_backend_async.rs:
```diff
@@ -0,0 +1,140 @@
+use utils::postgres_backend::{AuthType, Handler, PostgresBackend, QueryError};
+
+// generate client, server test streams
+async fn make_tcp_pair() -> (TcpStream, TcpStream) {
+    let listener = TcpListener::bind("127.0.0.1:0").await.unwrap();
+    let addr = listener.local_addr().unwrap();
+    let client_stream = TcpStream::connect(addr).await.unwrap();
+    let (server_stream, _) = listener.accept().await.unwrap();
+    (client_stream, server_stream)
+}
```
https://docs.rs/tokio/latest/tokio/io/fn.duplex.html

arssher replied:
Cool, have been searching for that and didn't find.

arssher replied:
... except that currently pgbackend is not parametrized with the stream. There is PR in progress by Arthur on that, we can change this there. And this is old code FWIW.

LizardWizzard replied:
Is it already changed there? If not lets add an issue with tech_debt tag. This also looks like a separate change that can be done in completely separate PR.

arssher replied:
I'd say that's the case when creating/managing the issue is much longer than the actual fix (once parametrization lands) and makes managing/triaging more important issues hard, but here you go.
https://github.com/neondatabase/neon/issues/3731

arssher commented on libs/pq_proto/src/framed.rs:
```diff
@@ -0,0 +1,265 @@
+}
+
+/// Write next message to the output buffer; doesn't flush.
+fn write_message(write_buf: &mut BytesMut, msg: &BeMessage<'_>) -> Result<(), ProtocolError> {
```
removed it

arssher commented on libs/pq_proto/src/framed.rs:
```diff
@@ -0,0 +1,265 @@
+}
+
+/// Flush out the buffer. This function is cancellation safe: it can be
+/// interrupted and flushing finished in the next call.
```
ok

arssher commented on libs/utils/Cargo.toml:
```diff
@@ -47,6 +47,8 @@ hex-literal.workspace = true
 criterion.workspace = true
 rustls-pemfile.workspace = true
+tokio-postgres.workspace = true
+tokio-postgres-rustls.workspace = true
```
Moved whole postgres_backend into separate crate. I dislike a bit 'lib.rs' filename in a single file crate though; also, I wonder whether many small crates increase compilation time due to linking the same dependencies multiple times incrementally.

LizardWizzard replied:
Its ok to have a single lib.rs in a crate. Bigger amount of smaller crates should increase parallelization and help with caching in incremental compilation

arssher commented on proxy/src/console/mgmt.rs:
```diff
@@ -33,61 +34,64 @@ pub fn notify(psql_session_id: &str, msg: ComputeReady) -> Result<(), waiters::N
-        let (socket, peer_addr) = listener.accept().await?;
+        let (socket, peer_addr) = listener
+            .accept()
+            .await
+            .context("failed to accept a new client")?;
```
removed

arssher commented on proxy/src/console/mgmt.rs:
```diff
@@ -33,61 +34,64 @@ pub fn notify(psql_session_id: &str, msg: ComputeReady) -> Result<(), waiters::N
-        socket
-            .set_nonblocking(false)
-            .context("failed to set client socket option")?;
+        let peer_addr = socket.peer_addr()?;
```
removed

arssher commented on libs/pq_proto/src/lib.rs:
```diff
@@ -191,255 +186,206 @@ pub struct FeExecuteMessage {
+    /// Read and parse startup message from the `buf` input buffer. It is
+    /// different from [`FeMessage::parse`] because startup messages don't have
+    /// message type byte; otherwise, its comments apply.
+    pub fn parse(buf: &mut BytesMut) -> Result<Option<FeMessage>, ProtocolError> {
```
Done that, please check again.

arssher commented on libs/pq_proto/src/lib.rs:
```diff
@@ -677,13 +634,12 @@ fn write_body<R>(buf: &mut BytesMut, f: impl FnOnce(&mut BytesMut) -> R) -> R {
-            io::ErrorKind::InvalidInput,
-            "string contains embedded null",
-        ));
+        return Err(ProtocolError::MessageParse(anyhow!(
+            "string contains embedded null"
+        )));
```
Probably, but this patch doesn't touch this, and it is already large. Do you think it is better do to it here, not separately?

arssher replied:
I went ahead and removed anyhow. However, I'm disappointed by the absense of .context() and ensure! analogs provided by thiserror, so personally I'd prefer to keep anyhow until this is not implemented.

arssher commented on libs/pq_proto/src/lib.rs:
```diff
@@ -731,7 +687,7 @@ impl<'a> BeMessage<'a> {
             BeMessage::AuthenticationSasl(msg) => {
                 buf.put_u8(b'R');
-                write_body(buf, |buf| {
+                write_body(buf, |buf| -> Result<(), ProtocolError> {
```
removed

arssher commented on libs/pq_proto/src/lib.rs:
```diff
@@ -191,255 +186,206 @@ pub struct FeExecuteMessage {
     #[error("Protocol error: {0}")]
     Protocol(String),
-    /// Failed to parse a protocol mesage
+    /// Failed to parse or, (unlikely), serialize a protocol message.
```
made it BadMessage

arssher commented on libs/utils/src/postgres_backend.rs:
```diff
@@ -93,48 +183,54 @@ impl fmt::Display for AuthType {
+enum MaybeWriteOnly {
+    Full(Framed<MaybeTlsStream>),
+    WriteOnly(FramedWriter<WriteHalf<MaybeTlsStream>>),
+    Broken, // temporary value palmed off during the split
```
Yes, it is not so easy. This is not the place where I want to fight with Rust.

arssher commented on libs/utils/src/postgres_backend.rs:
```diff
@@ -93,48 +183,54 @@ impl fmt::Display for AuthType {
+            MaybeWriteOnly::WriteOnly(_) => {
+                Err(io::Error::new(ErrorKind::Other, "reading from write only half").into())
+            }
+            MaybeWriteOnly::Broken => panic!("IO on invalid MaybeWriteOnly"),
```
I don't care much and I'll do that if you insist, however I actually find the current state slightly better in terms of readability. IO in `Broken` can happen only if there is an blunder in relatively compact piece of *internal* code inside postgres_backend. It just should never happen. `Broken` wouldn't even exist has rust been more intelligent in handing these replacement ; it should allow to temporary move out struct field if no error can happen before it is moved back. However, attempt to read from write only half is quite possible in case of external user error, and you might even do that sorta intentional in dirty code (like we had for a long time before this patch).

arssher replied:
citing [too many linked lists](https://rust-unofficial.github.io/too-many-lists/first-push.html) "In principle, this is something Rust could actually accept, but it won't (for various reasons -- the most serious being [exception safety](https://doc.rust-lang.org/nightly/nomicon/exception-safety.html))"

LizardWizzard replied:
I do not insist on removing broken variant. I only insist on choosing one strategy to handle invariant violation. I'm fine with either Result or panic. I dont see why we should use both to indicate the same thing.

I can see one possible direction on how Broken can be avoided, but as we agreed, it is not an easy way.

arssher replied:
I explained above why I intentionally chose to have two different strategies here, and why `WriteOnly` and `Broken` are different things. We can have a call if that wasn't clear.

arssher replied:
Hm, or you are talking about bail! in `Broken` state at `start_tls`? I'll replace it with panic.

LizardWizzard replied:
> Hm, or you are talking about bail! in Broken state at start_tls?

Exactly.

arssher replied:
fixed

LizardWizzard commented on safekeeper/src/receive_wal.rs:
```diff
@@ -2,204 +2,285 @@
+        pgb.write_message_flush(&BeMessage::CopyBothResponse)
+            .await?;
+
+        // Experiments confirm that doing network IO in one (this) thread and
```
Are there any references to these experiments?

arssher replied:
Added link to the related PR.

LizardWizzard replied:
The PR description only refers to batching reads from the socket. I see that disk IO operations are synchronous, so I can imagine how they can affect fully asynchronous network IO. Can it be that this is the reason why its better? I think its worth a comment update at least. 

Please correct me if I'm wrong.

As a side note, have you tried tokio's async disk io functions? It seems that if they're on par perf-wise this can significantly simplify the code here

arssher replied:
I expect win comes from two things:
1. calling fsync not on each arrived chunk
2. pipelining: doing network IO and disc IO in parallel. Async disc io won't replace that.

I don't immediately remember whether we had experiments differentiating these cases. Perhaps @petuhovskiy knows more.

LizardWizzard replied:
> pipelining: doing network IO and disc IO in parallel. Async disc io won't replace that.

Here to achieve this you need explicit separate thread. With async its hidden from you. And you can keep it parallel.

I asked about async read write rather out of curiosity because this approach is not mentioned anywhere.

LizardWizzard commented on safekeeper/src/receive_wal.rs:
```diff
@@ -2,204 +2,285 @@
+        // processing with disc IO in another significantly improves
+        // performance; we spawn off WalAcceptor thread for message processing
+        // to this end.
+        let (msg_tx, msg_rx) = channel(MSG_QUEUE_SIZE);
```
Not sure I understand why we need both split workaround and these channels. You can select on pgb.read_message() and reply_rx.recv(). This way you dont need split since you own the stream. If you received the message from the socket you forward it to separate thread. If you received reply message from WalAcceptor you write it to the same instance of pg backend. 

Am I misunderstanding it?

arssher replied:
Proposed schema is vulnerable to deadlock where both threads are blocked trying to send to the channel.

arssher replied:
try_send is not good, it should be retried, and when? busy loop or sleep? We anyway need to send at some point, and do that before attempting to get next message, otherwise the thing is pointless, you just accumulate messages in memory. You could imagine some state machine conditionally choosing what to do on each loop, but it doesn't look simpler.

> Additionally what do you think about using an unbounded channel? 

That's bad as memory consumption grows indefinitely. I don't think monitoring that with prometheus really solves this.

> E g [here](https://tomaka.medium.com/a-look-back-at-asynchronous-rust-d54d63934a1c) in paragraph "Flow control is hard" discusses the topic

Indeed flow control is hard. select! here specifically addresses the deadlock mentioned in that link.

LizardWizzard replied:
> try_send is not good,

If you cannot send you simply block on recv from reply channel as I explained above. When you received the message

> That's bad as memory consumption grows indefinitely. I don't think monitoring that with prometheus really solves this.

It will grow a bit, but shouldnt go really hi because you essentially it is a form of request response pattern.  If you cannot process requests fast enough you better know about it. And client shouldnt send too many data without receiving an ack from the other side.

> Indeed flow control is hard. select! here specifically addresses the deadlock mentioned in that link.

Select unnecessary overcomplicates control flow here because it has far richer semantics than try_join_all.

arssher replied:
> If you cannot send you simply block on recv from reply channel 

It can still be blocked after reply returns something. This is not a trivial thing, a complete description is needed to talk about that. Anyway, this is out of this patch scope as I and Arthur commented below.

> It will grow a bit, but shouldnt go really

Is is not generally bounded by anything, and this is not a request-response, it is bidirectional stream.

> Select unnecessary overcomplicates control flow here

`try_join_all` has different semantics not fitting here, as said elsewhere, we need `select!` exactly. Neither I generally consider `select!` being more complicated that `try_join_all`.

LizardWizzard replied:
>  Anyway, this is out of this patch scope as I and Arthur commented below.

As I get it this change is the main motivation for the patch. It is the motivation for all the squats around split and Framed. So IMO it is exactly in scope of this patch. 

> It can still be blocked after reply returns something

And thats fine, try_send it again on next iteration. 

> try_join_all has different semantics not fitting here

Replied to the relevant thread

>  request-response, it is bidirectional stream.

It is, if I understand it correctly the stream is a essentially a request response with the difference that a bunch or requests can be batched, right? Can there be a huge amount of messages without reply from the other side? From my understanding no, but I may be wrong

arssher replied:
On which iteration? I don't follow. Should we select in the meanwhile on reading more data? If yes, we pointlessly and unboundedly accumulate data in memory, as I said. If no, it means complicated state machine. And how we read more data without split if reply (out) buffer can be also full, i.e. flushing is WOULDBLOCK? As I said, full description is required to talk about that. Generally, it is not hard to observe that in arbitrary bidi stream (and this is the case here) at least one side must poll reading and writing concurrently, otherwise deadlock risk is inevitable. And I'd prefer to not make any assumptions on how the other side behaves, so better make *both* sides do concurrent read writes.

arssher replied:
> Can there be a huge amount of messages without reply from the other side? 

Yes, there is generally no constraints. And adding them only to justify poorly written protocol implementation doesn't sound good.

arssher commented on libs/utils/src/postgres_backend.rs:
```diff
@@ -49,20 +89,70 @@ pub trait Handler {
 #[derive(Clone, Copy, PartialEq, Eq, PartialOrd)]
 pub enum ProtoState {
     Initialization,
+    // Encryption handshake is done; waiting for encrypted Startup message.
```
added

arssher commented on libs/utils/src/postgres_backend.rs:
```diff
@@ -154,210 +250,294 @@ pub fn query_from_cstring(query_string: Bytes) -> Vec<u8> {
-        }
-        .map_err(QueryError::from)
+    /// Polling version of read_message, saves the caller need to pin.
+    pub fn poll_read_message(
```
It is a leftover from experiments in https://github.com/neondatabase/neon/pull/3522
I'd leave it, it is simple and might be useful in the future.

LizardWizzard replied:
Personally I dont agree with that argument. Dead code is dead code. This function is three lines long and can be added when it is actually starts being used.

If you still insist on keeping it please add a comment and a corresponding unit test.

arssher replied:
removed

arssher commented on libs/utils/src/postgres_backend.rs:
```diff
@@ -154,210 +250,294 @@ pub fn query_from_cstring(query_string: Bytes) -> Vec<u8> {
+        S: Future,
+    {
+        let ret = self.run_message_loop(handler, shutdown_watcher).await;
+        self.framed.shutdown().await.ok();
```
removed swallowing

arssher commented on libs/utils/src/postgres_backend.rs:
```diff
@@ -435,7 +641,7 @@ impl PostgresBackend {
 
             FeMessage::Describe(_) => {
-                self.write_message_noflush(&BeMessage::ParameterDescription)?
+                self.write_message(&BeMessage::ParameterDescription)?
```
> Should the second one change to write_message_flush?

No, there is flush after `process_message`.

> Personally I'm against the naming change. 

https://github.com/neondatabase/neon/pull/3576#issuecomment-1429434944

> Do we have tests for this code? If not IMO we need to add them as part of this PR.

Unfortunately I can't immediately say, but I'm sure that semantics of this code hasn't changed, and this PR is already huge, so I don't want to touch this.

LizardWizzard replied:
> No, there is flush after process_message.

Hm, taking this into account it seems that sometimes we may call flush twice. First time in handler and second time in process_message. E g this is the case in [handle_identify_system](https://github.com/neondatabase/neon/blob/ad413846ed7044f3e7a59ffe80e0fd89b38ff0bb/safekeeper/src/handler.rs#L274).

It is not a big problem in any sense, but not definitely not a good thing either. If it was this way before this is definitely out of scope.

> Personally I'm against the naming change.
> https://github.com/neondatabase/neon/pull/3576#issuecomment-1429434944

Ok, thats subjective, I wont insist

funbringer replied:
I'm also against the naming change. I especially appreciate the explicit `_noflush` prefix. https://github.com/neondatabase/neon/pull/3576#issuecomment-1429434944 didn't convince me. This is a subtle and error-prone change, it invalidates team assumptions and is thus harmful.

Furthermore, proxy, which sticks with the `_noflush`, also uses `pq_proto`, and I wouldn't want to mix write/write_noflush & write_flush/write in the same code (which you effectively impose [here](https://github.com/neondatabase/neon/pull/3576/files#diff-ad02c0cf4770ea1daad69234627b8beb1ad4f78017395917eab6a00c55931f8fR76-R81)).

It's the changes like this one which make it harder to believe that the PR's size is justified.

arssher replied:
> Hm, taking this into account it seems that sometimes we may call flush twice. First time in handler and second time in process_message. E g this is the case in [handle_identify_system](https://github.com/neondatabase/neon/blob/ad413846ed7044f3e7a59ffe80e0fd89b38ff0bb/safekeeper/src/handler.rs#L274). It is not a big problem in any sense, but definitely not a good thing either. If it was this way before this is definitely out of scope.

Will fix, it is reasonable to just flush after each process_message as it is done currently in postgres_backend_async.

arssher commented on libs/utils/src/postgres_backend.rs:
```diff
@@ -475,11 +681,209 @@ impl PostgresBackend {
+
+    /// Log as info/error result of handling COPY stream and send back
+    /// ErrorResponse if that makes sense. Shutdown the stream if we got
+    /// Terminate. TODO: transition into waiting for Sync msg if we initiate the
```
No. We never do anything after COPY stream ends, and there are no plans/needs to do that in the future. TODO is more about strictly following the protocol.

arssher commented on libs/utils/src/postgres_backend.rs:
```diff
@@ -475,11 +681,209 @@ impl PostgresBackend {
+
+        let err_to_send = match &end {
+            ServerInitiated(_) | Other(_) => Some(end.to_string()),
+            // Note: we should probably close the socket, as
```
I reworded the comment.

arssher replied:
There is already one on this subject
https://github.com/neondatabase/neon/issues/458
I'd say that likely the proper way would be to 1) rework replication patch for tokio-postgres to not send CopyFail 2) fix https://github.com/sfackler/rust-postgres/issues/755

arssher commented on libs/utils/src/postgres_backend.rs:
```diff
@@ -475,11 +681,209 @@ impl PostgresBackend {
+    ) -> Poll<Result<usize, std::io::Error>> {
+        let this = self.get_mut();
+
+        // It's not strictly required to flush between each message, but makes it easier
```
The PR doesn't change this place. Answering the question, it depends on what the user does. This is used only for sending basebackup, and it indeed as comment says it looks decent in this sense.

arssher commented on safekeeper/src/handler.rs:
```diff
@@ -62,11 +60,20 @@ fn parse_cmd(cmd: &str) -> anyhow::Result<SafekeeperPostgresCommand> {
         Ok(SafekeeperPostgresCommand::JSONCtrl {
             cmd: serde_json::from_str(cmd)?,
         })
+    } else if cmd_lowercase.starts_with("show") {
```
Yes, as comments say this is for `pg_receivewal`. No, I'm fine having it here. Will add to the commit message.

LizardWizzard replied:
IMO the patch is already big enough to not add additional unrelated changes on top

arssher replied:
It is not invasive and not critical.

LizardWizzard replied:
This is not an argument for me. The change is logically independent. This time it is this change, next PR will contain other ones. This shouldnt be a way how we as a team approach review workflow.

Frankly speaking I'm quite tired of big PR's that mix possibly independent changes. Please look at this from a reviewer perspective. Hope for your understanding.

arssher replied:
Ok, I extracted it into separate commit, should be faster than discussing.

arssher commented:
AFAIS I addressed all reviews so far, please have another look.

LizardWizzard reviewed (changes requested):
I took another look. I'm still not on board with the changes. 

Additionally, this patch contains unrelated changes (renaming functions, new psql show command). I would recommend splitting them out because the patch is really big and its hard to keep track of everything during reviews.

I'm still thinking that custom implementation of Framed that is slightly different from one in tokio is not what we need here.

The main motivation for the patch seems to be this [change](https://github.com/neondatabase/neon/pull/3576/files#diff-151f3c95a082f4bd3749306b81d777fa4d2f598b938221c9d5157aaed75fb912R64) in safekeeper. First of all the PR title and description do not mention this. Commit messages do not reflect this intention and other changes included in this patch. As I said in one of the comments IMO the PR would benefit from splitting into independent parts each with its own motivation. Mixing independent changes leads to extended review time because of increased patch size and at some point many discussions start to block code that has no concerns to be merged without waiting for other conversations to resolve. 

Secondly, to my mind similar result can be achieved without requiring split machinery (either tokio or futures one). Both of them seem to have their own cons, allocation of waker in futures, thread::yield_now() in tokio and in general this concept of split contains pitfalls exactly around use case with polling from different tasks (it is [mentioned](https://github.com/neondatabase/neon/blob/550ef5bd3f2bfe10ba0daa2db108dd7e039de58d/libs/pq_proto/src/framed.rs#L120) in your comments too). To me better way to achieve desired goal would be to change the way how these two sk tasks communicate with each other. See https://github.com/neondatabase/neon/pull/3576/files#r1118916294 and https://github.com/neondatabase/neon/pull/3576#discussion_r1118917633

Additionally taking into account https://github.com/neondatabase/neon/issues/3667 I'm really not a fan of such a big refactoring without good unit test coverage. This refactoring is an opportunity to move to better structure for our whole postgres protocol handling machinery. Since we deal with parsing of data that came through network it makes sense to add some fuzzying routines. With all the rust tooling available it shouldnt be a really big task.

petuhovskiy commented:
> Additionally, this patch contains unrelated changes (renaming functions)

If you are talking about `write_message_noflush -> write_message`, it actually was not renamed in this PR, it's a strange looking diff coming from migrating to `postgres_backend_async` from `postgres_backend`. The thing is, before this PR, we had:

- [postgres_backend](https://github.com/neondatabase/neon/blob/000eb1b069ce5f17b8711130488f1fe8bd206749/libs/utils/src/postgres_backend.rs#L242-L261) (sync) with functions:
  * write_message_noflush
  * flush
  * write_message (write with flush)
- [postgres_backend_async](https://github.com/neondatabase/neon/blob/000eb1b069ce5f17b8711130488f1fe8bd206749/libs/utils/src/postgres_backend_async.rs#L225-L239) with functions:
  * flush
  * write_message (write without flush)

This PR added `write_message_flush` to the async version and replaced postgres_backend with postgres_backend_async. That's why it's looking like a rename, but in fact we already had `write_message()` with different meaning between sync and async versions.



> I would recommend splitting them out because the patch is really big and its hard to keep track of everything during reviews.

Yes, it's easier to review several small patches instead of one big, but it takes more time to prepare several small independent PRs.

> The main motivation for the patch seems to be this [change](https://github.com/neondatabase/neon/pull/3576/files#diff-151f3c95a082f4bd3749306b81d777fa4d2f598b938221c9d5157aaed75fb912R64) in safekeeper.

If you are talking about `let (msg_tx, msg_rx) = channel(MSG_QUEUE_SIZE);`, we already had similar logic for a long time. https://github.com/neondatabase/neon/blob/000eb1b069ce5f17b8711130488f1fe8bd206749/safekeeper/src/receive_wal.rs#L146-L205

This PR mostly preserves it, but also adds a buffered channel for responses, which is now easy to do because of async tasks.

I'd say that this PR does mostly what is says, i.e. Removes synchronous postgres_backend; Adds framed.rs; Fixes CopyBoth errors in the logs. The diff looks large because a lot of code was moved in the process, but the logic stays the same in the most places.

LizardWizzard commented:
> If you are talking about let (msg_tx, msg_rx) = channel(MSG_QUEUE_SIZE);, we already had similar logic for a long time.

I'm talking about the whole idea around split and the way how different task is used rather than this exact line.

> Yes, it's easier to review several small patches instead of one big, but it takes more time to prepare several small independent PRs.

I think you're trying to optimize the wrong part here, for the person who submits the patch its faster to prepare a bigger one, but at the end of the day merging smaller patches is faster than going through reviews for one big PR. Additionally for bigger patches it is easier to miss bugs just because of the amount of code and increased number of iterations. So IMO optimizing for review speed and thus total amount of time it takes to deliver a feature is a better strategy. We should optimize for better work as a team, not for individual contributor convenience.

> If you are talking about write_message_noflush -> write_message,

Not only this, another example is pg_receivewal support. 

Regarding the rename.  It looks like more places used sync semantics, and you suggested to rename them in first comment:

> write_message_noflush -> write_message and write_message -> write_message_flush can be confusing or error-prone, inverting the write_message semantics looks strange

Which lead to increased diff.

Personally I dont see a reason why this needs to be done in this patch. It can be split out to separate one that merges naming between sync and async.

Also I do think that write_message -> serialize_message can completely side step the confusion.

LizardWizzard commented:
> Yes, you are proposing to remove split and write a slightly different code, 

I propose avoiding the necessity to write extra code. To me custom framed implementation is completely unnecessary.

> the current version is ported from the sync code that is known to work well.

Code in main doesnt rely on tokio::io::split. I dont see why this is an argument here.

> I haven't suggested to rename anything
> inverting the write_message semantics looks strange

For code that was previously sync it is exactly a rename and this patch inverts the semantics. Additionally see https://github.com/neondatabase/neon/pull/3576#discussion_r1119126068

>  it will require renaming pageserver code that is already using postgres_backend_async and it leads to more diff.

This is why the entire naming conversion thing should be a separate patch.

arssher commented:
> I'm going to do switch to write_message_noflush/write_message in the whole project. I can extract that to separate commit, would that help?

Did that, doesn't help much with diff...
And again, all review is addressed so far.

arssher commented:
Rebased again.

LizardWizzard commented:
> Did that, doesn't help much with diff...

It helped a bit. There is a separate commit with this change now, so the second commit became a little smaller. If you split more changes I believe it will shrink a bit more.

> It probably technically can, but it is not obvious at all that the result would be [simpler/better](https://github.com/neondatabase/neon/pull/3576/files#r1121721992).

I dont understand this as an argument.

> Exactly same problems arise in safekeeper walsender, solving that at pgbackend level avoids doing that twice.

Could you elaborate? I may lack some context

> consequences of fighting with Rust strictness.

This is only partially true. Rust can be annoying and sometimes the suitable approach significantly differs from what you can think ahead of time. As I previously said in the comment, is there a reason why you cannot hold owned stream without using split and dispatch messages in it? 

Rust strictness is about guarantees. If you cannot guarantee that it would be the same task then you should consider the worst case. Can someone misuse split abstraction and write code that starts polling it from multiple tasks? Sure. Can we do better? I think so, and refactoring IMO is a perfect place to explore the opportunity.

> However, neither I view that as highest priority here;

Ok, when this can be a priority other than in refactoring PR that tries to reduce tech debt? 

> but it is simple and gets tested in integration tests

Integration tests have their weaknesses. In motivation paragraph here https://github.com/neondatabase/neon/issues/3667 I explain why having unit tests increases developer velocity. This is true especially for small independent components like pg backend. Is it that hard to test Copy that we need to add follow up issue for that?

Additionally being "simple" and "internal" cannot be a reason why this shouldnt be covered.

You changed a whole lot of code here. And as I already mentioned this PR tries to be a refactoring.

> Doing that in other way (not with split) is very loosely related work

I disagree. The refactoring you did tries hard to mirror logic from previous sync version. Significant part of code is dedicated to this job. So this is exactly about the code added in this patch, thus perfectly relevant.

> Of course, because it is sync. It instead impressively relies on our own crate specifically designed for splitting potentially tls enabled sync streams.

So we fully rewrite this code for this logic to remain the same. This refactoring is an opportunity to make it clearer and do not return back to this code. Imagine we merged the PR. Tech debt associated with this is still there. Someone needs to get back to it and rewrite it again. Why do we need to rewrite the same thing twice if we already know which problems exist in newer version? 

> So if someone is going for fuzzying that's okay, but not in this PR.

Sure, I'm fine with it. Please create corresponding issue.

> Currently this patch blocks other work, e.g. Arthur's measuring traffic PR, that's really bad

IMO this PR in its current form moves forward with uniting everything under async umbrella but it still keeps the tech debt that was there but in a bit different form despite the fact that it is a significant rewrite of the component. So if your proposal is to leave it here, merge the PR and be done, I'm not on board with the approach. Sorry. 

Maybe you need someone else to break the knot.

kelvich commented:
Some thoughts after reading this thread.

* avoiding allocation in the hot code path is IMO a good reason to not to use tokio Framed. In the case of safekeeper where most of network messages end up being written on disk it is not that important (as fsync will be order of magnitudes slower anyway). But this parsing code is used in read only requests (e.g. in pageserver and proxy) as well, so it is important. I agree that there are a good reasons to stick to the standard functions, but here I'm leaning towards the approach proposed in the PR.
* current way of socket reading and disk writing in the safekeeper emerged after a big series of tests that took quite a long time and improved safekeeper performance about 5 times and finally surpassed speed of postgres replication. A lot of this speed came from emergent phenomenons (how much messages are piling up in the queue, what is the average size of the message, what is the fsync latency on our hardware, etc) that contribute to avoiding stalls in the walsender->walreceiver->walwriter pipeline. I think that it is some context that Arseny and Arthur had and reviewers did not, hence some amount of misunderstanding (however Arseny explained this in the comments). So +1 to the way it is done in the PR and -1 to the suggested changes until they are extensively perf-tested.
* I get Arseny's explanation of why `try_send` will not work without significantly complicating state machine
* I see that the current approach is labeled as a tech debt, but don't see any arguments on why. May be it is, but just by  looking at the comments it seems like a matter of personal preference.

So to sum it up I'm +1 to proceed with the current PR. I also not against trying to avoid sock_split if it will simplify things, but better as a separate PR to check for any perf impact.

arssher commented:
Just a minor correction:
> avoiding allocation in the hot code path is IMO a good reason to not to use tokio Framed.

There are two 'allocation during each message processing' issues. First is in [bilock.rs](https://github.com/neondatabase/neon/pull/3576#discussion_r1113469372) which is used when you split tokio Framed. In non-split usage it doesn't exist, so pageserver and proxy would be ok with that. Another one in our [message reading code](https://github.com/neondatabase/neon/blob/a60f687ce2b9022f10943b537daf05b09a9d48fc/libs/pq_proto/src/lib.rs#L294) , which affects everything. Switch to framed, whether custom or tokio removes that.

LizardWizzard commented:
@kelvich 
> avoiding allocation in the hot code path is IMO a good reason to not to use tokio Framed.

As Arseny pointed out this is not relevant to tokio framed. Instead with current usage of split there is a "spin lock" with thread::yield_now() on AtomicBool. See https://github.com/tokio-rs/tokio/blob/3d33610ed2420111982e5a42c764761c9060e6ab/tokio/src/io/split.rs#L147. Do we know for sure that this does not have any negative performance impact?

> allocation in the hot code path is IMO

Do we know for sure that this allocation has impact on performance? I agree that this is not good if we can easily avoid it. But if it doesnt have perf impact then having simpler code or less amount of it looks like a better option to me.

> I think that it is some context that Arseny and Arthur had and reviewers did not, hence some amount of misunderstanding (however Arseny explained this in the comments). So +1 to the way it is done in the PR and -1 to the suggested changes until they are extensively perf-tested.

1. If the context is only in people's heads this is also a problem. Important bits should be reflected in documentation and clear references should be given.
2. The suggested approach shouldnt be different from what Arseny described as an intended solution. Regarding the state machine complexity, we (myself and @funbringer) provided a combinator that is abstracted over `AsyncRead + AsyncWrite` and protocol implementation in form of `tokio_util::codec::{Encoder, Decoder}`. Standalone example can be found here:  https://github.com/LizardWizzard/duplex/blob/master/src/main.rs#L43. I posted it with description in our slack channel where we've discussed the PR.

>  until they are extensively perf-tested

Do we have a test case in our test suite that can be used for validation? Did we check our usual perf numbers for this patch to not cause any perf degradation?

> I see that the current approach is labeled as a tech debt, but don't see any arguments on why.

IMO arguments are clear:

1. The well known problem with split: https://github.com/tokio-rs/tls/issues/40. There is even a comment in our code that mentions this issue. The answer to it was that we dont do it in multiple tasks, but I dont want to be the person accidentally spawning a task and waste time on debugging and ending up either working around spawn, or with another refactoring PR that removes dependency on split. If you think that I'm too paranoid on that, lets ask somebody else, it may be indeed the case.
2. IMO reimplementing something that does precisely what tokio's Framed and tokio_util::codec abstractions were designed for is not a good practice.

If these still are not arguments for you, ok. Please note that this PR doesnt have an approval from storage team member (pageserver side of things). Codeowner-wise IMO it is important that each involved team gives an approval on a patch that touches core parts of the code base that the team uses to build components it is responsible for. So IMO thats reasonable to ask someone else. I would suggest Joonas or Christian.

And I havent seen an approval from @funbringer either.

***

Anyway, thanks @arssher for fixing all the comments

LizardWizzard commented:
Can we close this?

arssher commented:
Superseded by https://github.com/neondatabase/neon/pull/3759
